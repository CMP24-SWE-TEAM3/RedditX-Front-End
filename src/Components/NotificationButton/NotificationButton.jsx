// imports
import {IoMdNotificationsOutline} from 'react-icons/io';
import {GlobalButtonStyled} from "Components/GlobalButton/GlobalButton.styled";
import React, {useState, useRef, useEffect} from 'react';
import Overlay from 'react-bootstrap/Overlay';
import Popover from 'react-bootstrap/Popover';
import {Container} from "./NotificationButton.styled";
import {BsThreeDots, BsClipboardCheck} from 'react-icons/bs';
import {FiSettings} from 'react-icons/fi';
import link from 'Assets/Images/link.png';
import db from 'Data/db.json';
import OverlayTrigger from "react-bootstrap/OverlayTrigger";
import Tooltip from "react-bootstrap/Tooltip";
import {Link} from 'react-router-dom';

/**
 * Component that displays notifications about changes in the state of the application
 *
 * @returns {React.Component}
 */
const NotificationButton = () => {

    /**
     * state of the notification button is changed  when the user clicks on the notification button
     */
    const [show, setShow] = useState(false);

    /**
     * state to track if mouse clicks outside notification's button or not
     */
    const [target, setTarget] = useState(null);

    // ref the event listener
    const ref = useRef(null);
    const wrapperRef = useRef(null);

    /**
     * Function that is called when the state of the application changes (when clicks outside of notification's buttons)
     * @param ref
     */
    const useOutsideAlerter = (ref) => {
        useEffect(() => {

            //Close toast if clicked on outside of element
            const handleClickOutside = (event) => {
                const box = document.getElementById("notification-box");
                if (ref.current && !ref.current.contains(event.target) && box!==event.target && !box.contains(event.target)) {
                    setShow(false);
                }
            }

            // Bind the event listener
            document.addEventListener("mousedown", handleClickOutside);
            return () => {
                // Unbind the event listener on clean up
                document.removeEventListener("mousedown", handleClickOutside);
            };
        }, [ref]);
    }

    /**
     * Function hand clicks  on the button notification
     * @param event
     */
    const handleClick = (event) => {
        setTarget(event.target);
        setShow(!show);

    };
    useOutsideAlerter(wrapperRef);

    return (
        <Container ref={ref}>
            <OverlayTrigger
                key={'bottom'}
                placement={'bottom'}
                overlay={
                    <Tooltip id={`tooltip-bottom`}>Notification</Tooltip>}>
                <GlobalButtonStyled id={'notification-box'} onClick={handleClick}>
                    <IoMdNotificationsOutline/>
                </GlobalButtonStyled>
            </OverlayTrigger>
            <Overlay
                show={show}
                target={target}
                key={Math.random()}
                placement="bottom"
                container={ref}
                ref={wrapperRef}>
                <Popover id="popover-contained">
                    <Popover.Header>
                        <span className={'word'}>
                            Notifications
                        </span>
                        <span className={'right-side'}>
                            <span className={'message'}>
                                <Link href={'https://www.reddit.com/message/messages'}>Messages </Link>
                            </span>

                            <button className={'but'}>
                                <i><BsClipboardCheck/></i>
                            </button>
                            <Link className={'setting'} href={"https://www.reddit.com/settings/notifications"}>
                                <i>
                                <FiSettings/>
                                </i>
                            </Link>
                        </span>
                    </Popover.Header>
                    <Popover.Body>
                        <div className={'pop'}/>
                        <div className={'content'}>
                            {db.comments.map(comment => {
                                return (
                                    <i className={'content-i'} key={Math.random()}>
                                        <Link className={'content-a'}
                                              href={'https://www.reddit.com/r/Eln2aa4yn/comments/yi1rd4/ccccc/'}>
                                    <span className={'a-span'}>
                                     <img src={link} alt={'community-name'}/>
                                    </span>
                                            <span className={'span-info'}>
                                        <span className={'child'}>
                                            <span className={'sub-child'}>
                                            <span className={'info'} key={comment.id}>
                                                Now in {comment.name}
                                            </span>
                                            <span className={'dot'}>.</span>
                                            <span className={'time'}>{comment.time}h</span>
                                            </span>
                                            <button className={'dot'}>
                                                <i><BsThreeDots/></i>
                                            </button>
                                        </span>
                                        <span className={'info-child'}>{comment.message}</span>
                                    </span>
                                        </Link>

                                    </i>)
                            })}

                        </div>
                    </Popover.Body>

                    <footer>
                        <Link href={'https://www.reddit.com/notifications'}>See All</Link>
                    </footer>
                </Popover>
            </Overlay>
        </Container>
    );
}

export default NotificationButton;
