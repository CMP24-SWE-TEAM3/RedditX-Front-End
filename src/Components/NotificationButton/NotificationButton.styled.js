import styled from "styled-components";

export const Container = styled.div`
  //background-color: ${({theme}) => theme.background.primary};
  color: ${({theme}) => theme.color.muted};


  .popover {
    background-color: ${({theme}) => theme.background.primary};
    border-radius: 4px;
    box-shadow: 0 4px 4px rgb(0 0 0 / 25%);
    //margin-top: 8px;
    overflow-y: unset;
    overflow: unset;
    width: 375px;
    max-width: 375px;
    // z-index: 100;
  }

  & .word {
    font-size: 14px;
    font-weight: 500;
    line-height: 18px;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    margin-right: 12px;
  }

  & .right-side {
    align-items: center;
    display: flex;
    overflow: hidden;
  }

  & .popover-header {
    background-color: ${({theme}) => theme.background.primary};
    border-radius: 4px;
    display: flex;
    justify-content: space-between;
    padding: 16px;
    border-bottom: none !important;
  }

  & .message {
    border-right: 1px solid ${({theme}) => theme.background.primary};
    padding-right: 8px;

    a {
      text-decoration: none;
      font-family: Noto Sans, Arial, sans-serif;
      font-size: 12px;
      line-height: 16px;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
      font-weight: 600;
      color: ${({theme}) => theme.color.gray};

      :hover {
        text-decoration: none !important;
      }
    }
  }

  & button {
    //margin-left: 8px;
    background: transparent;
    border: none;
    color: inherit;
    cursor: pointer;
    padding: initial;

  }

  i {
    color: ${({theme}) => theme.color.gray};
    font-size: 20px;
    font-weight: 400;
    height: 20px;
    line-height: 20px;
    vertical-align: middle;
    width: 20px;
  }

  .setting {
    margin-left: 8px;
  }


  // body of notification
  .popover-body {
    //height: 77px;
    max-height: 250px;
    background-color: ${({theme}) => theme.background.primary};
    overflow: hidden;
    overflow-y: auto;
    padding: 0;
    transition: max-height .3s ease-out;
  }

  .pop {
    margin: 0 16px;
  }

  .content-i {
    margin: 0;
    padding: 0;
    border: 0;
    font: inherit;
    vertical-align: baseline;
  }

  .content {
    position: relative;


    img {
      border-radius: 50%;
      height: 32px;
      width: 32px;
      background-color: ${({theme}) => theme.background.hover_background};
      display: block;
      object-fit: cover;
    }
  }

  .content-a {
    display: flex;
    padding: 16px;
    text-decoration: none;
  }

  .span-info {
    flex: 1 0;
  }

  .a-span {
    padding-right: 8px;
    position: relative;
  }

  .child {
    align-items: center;
    display: flex;
    justify-content: space-between;
    margin-bottom: 4px;
  }

  .sub-child {
    margin: 0;
    padding: 0;
    border: 0;
    font: inherit;
    vertical-align: baseline;
  }

  .info {
    font-size: 14px;
    line-height: 21px;
    font-family: Noto Sans, Arial, sans-serif;
    font-weight: 400;
    color: ${({theme}) => theme.color.primary};
  }

  .dot {
    font-size: 12px;
    font-weight: 400;
    margin: 0 5px;
    line-height: 16px;
    color: ${({theme}) => theme.color.gray};

    :hover {
      background-color: ${({theme}) => theme.background.post_background};
    }
  }

  .time {
    font-family: Noto Sans, Arial, sans-serif;
    font-size: 14px;
    font-weight: 400;
    color: ${({theme}) => theme.color.gray};
    line-height: 19px;
  }

  .info-child {
    font-family: Noto Sans, Arial, sans-serif;
    font-weight: 400;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: normal;
    -webkit-box-orient: vertical;
    font-size: 14px;
    color: ${({theme}) => theme.color.gray};
    word-break: break-word;
    line-height: 18px;
    max-height: 54px;
    -webkit-line-clamp: 3;
  }

  // footer styles
  footer {
    align-items: center;
    background-color: ${({theme}) => theme.background.post_background};
    border-radius: 0 0 4px 4px;
    display: flex;
    height: 49px;
    justify-content: center;
    padding: 0 12px;

    a {
      font-size: 14px;
      font-weight: 700;
      letter-spacing: .5px;
      line-height: 32px;
      text-decoration: none;
      text-transform: uppercase;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
      color: #24a0ed;
    }
  }
`