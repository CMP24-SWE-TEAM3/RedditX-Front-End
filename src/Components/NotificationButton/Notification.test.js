import {shallow} from 'enzyme';

// import Components
import NotificationButton from "./NotificationButton";

describe("Notification button Component", () => {
    it("this is a test for Notification button Component", () => {
        expect(shallow(<NotificationButton/>)).toMatchSnapshot();

    });
});