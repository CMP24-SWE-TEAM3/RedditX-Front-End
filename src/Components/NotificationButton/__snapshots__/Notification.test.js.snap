// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification button Component this is a test for Notification button Component 1`] = `ShallowWrapper {}`;
